README.md
# Mobile Payments

An easy way to integrate mobile payments into your web project.

## Motivation

Recently, VodaCom has released open API portal for M-Pesa. The following are some issues arose:

For Python integration, they have provided the package in zipped format. It is not user friendly and some developers found it challenging in installation.

Also, the sample code provided for Python implementation are long and most comprise repitition and become prone to errors for developers.

Mobile Payments package try to resolve above issues for Python by providing easy and user friendly installation, also the refactored code that is simplified.

## Prerequisites

Python 3.6+
Pip

## Installation

This package is available in [Python Package Index](https://pypi.org/project/mobile-payments/) and can be installed using `pip`

```
pip install mobile-payments
```

The package comprise both original open API codes and refactored codes.

To use original open API code import `open_api` module

```
from mobile_payments.open_api import APIContext, APIMethodType, APIRequest
```

To use refactored code import `MPESA` from `vodacom` module.

```
from mobile_payments.vodacom import MPESA
```

## Features

Ready methods

- [x] Customer-to-business
- [x] Business-to-customer
- [x] Business-to-business
- [x] Reversal
- [x] Transaction status

## Requisites

The following are required and are obtained from [Vodacom Open Api portal](https://openapiportal.m-pesa.com/login)

- Api Key
- Public Key

## Examples

The following examples requires `api key`. It can be obtained by registering to [open api portal](https://openapiportal.m-pesa.com) and follow the instructions.

It also uses refactored code, if you want to use original open API code it is also possible but follow the examples in documentation.

### Customer to Business payment via vodacom m-pesa

```python
# vodacom M-PESA
from mobile_payments.vodacom import MPESA

api_key = '<your-api-key>'
public_key = '<open-api-public-key>'

m_pesa = MPESA(api_key=api_key, public_key=public_key)

# Customer to Business payment
parameters = {
    'input_Amount': 10, # amount to be charged
    'input_Country': 'TZN',
    'input_Currency': 'TZS',
    'input_CustomerMSISDN': '000000000001',
    'input_ServiceProviderCode': '000000',
    'input_ThirdPartyConversationID': 'c9e794e10c63479992a8b08703abeea36',
    'input_TransactionReference': 'T23434ZE3',
    'input_PurchasedItemsDesc': 'Shoes',
}

response = m_pesa.customer2business(parameters)
```

Sample response

```
{'body': {'output_ResponseCode': 'INS-0',
'output_ResponseDesc': 'Request processed successfully',
'output_TransactionID': '79eKKNrYVfCj',
'output_ConversationID': 'c9e794e10c63479992a8b08703abeea36', 'output_ThirdPartyConversationID': 'asv02e5958774f7ba228d83d0d689761'}}
```

## Integrating in Django

If doing web project using Django framework

```python
# views.py
from mobile_payments.vodacom import MPESA

def payments(request):
    if request.method == 'POST':
        form = PaymentForm(request.POST)
        if form.is_valid():
            m_pesa = MPESA(api_key, public_key)

            parameters = {
                'input_Amount': 10,
                'input_Country': 'TZN',
                'input_Currency': 'TZS',
                # '000000000001' - phone number for testing,
                'input_CustomerMSISDN': request.POST.get('phone'),
                'input_ServiceProviderCode': '000000',
                'input_ThirdPartyConversationID': get_random_string(18),
                'input_TransactionReference': get_random_string(18),
                'input_PurchasedItemsDesc': 'Shoes',
            }

            results = m_pesa.customer2business(parameters)

            if results.body['output_ResponseCode'] == 'INS-0':
                # successful transaction

            else:
                # unsuccessful transaction

```

More details on response codes, see [api documentation](https://openapiportal.m-pesa.com/api-documentation)

## License

Code released under [MIT LICENSE](https://github.com/ZendaInnocent/mobile-payments/blob/main/LICENSE)


CHANGELOG.md
# Latest Version

- Change `query_transaction_status` method to `status`.


tests/test_vodacom_mpesa.py
from unittest.mock import patch

from mobile_payments.constants import *
from mobile_payments.vodacom import MPESA


def test_get_encrypted_api_key():
    m_pesa = MPESA(api_key=MPESA_API_KEY, public_key=MPESA_PUBLIC_KEY)

    enc_key = m_pesa.get_encrypted_api_key()

    assert enc_key is not None


@patch('mobile_payments.vodacom.MPESA.get_session_id')
def test_get_session_id(mock_get_session_id):
    mock_get_session_id.return_value = 'some random key'

    m_pesa = MPESA(api_key=MPESA_API_KEY, public_key=MPESA_PUBLIC_KEY)

    session_id = m_pesa.get_session_id()

    assert session_id == 'some random key'


tests/test_mpesa_reversal.py
"""Test suite for testing VodaCom M-Pesa reversal transaction
Testing common responses:
- successful transaction: status code 201
- duplicate transaction: status code 409
"""

from unittest.mock import patch

from mobile_payments.defaults import *
from mobile_payments.vodacom import MPESA

from .samples import (duplicate_transaction_sample, reversal_parameters,
                      successful_transaction_sample)


class TestMpesaReversalTransaction:
    @classmethod
    def setup(cls):
        cls.m_pesa = MPESA(api_key=MPESA_API_KEY, public_key=MPESA_PUBLIC_KEY)
        cls.mock_reversal_patcher = patch(
            'mobile_payments.vodacom.MPESA.reversal')
        cls.mock_reversal = cls.mock_reversal_patcher.start()

    @classmethod
    def teardown_class(cls):
        cls.mock_reversal_patcher.stop()

    def test_reversal_successful_transaction(self):
        """Test for successful reversal transaction
        Status code should be 201
        """
        self.mock_reversal.return_value.json.return_value = \
            successful_transaction_sample

        response = self.m_pesa.reversal(reversal_parameters)

        assert self.mock_reversal.called
        assert response.json() == successful_transaction_sample
        assert response.json()[0]['status_code'] == 201

    def test_reversal_duplicate_transaction(self):
        """Test for duplicate reversal transaction
        Status code should be 409
        """
        self.mock_reversal.return_value.json.return_value = \
            duplicate_transaction_sample

        response = self.m_pesa.reversal(reversal_parameters)

        assert self.mock_reversal.called
        assert response.json() == duplicate_transaction_sample
        assert response.json()[0]['status_code'] == 409


tests/test_mobile_payments.py
from mobile_payments import __version__


def test_version():
    assert __version__ == '0.1.2'


mobile_payments/open_api.py
from enum import Enum
import json

import requests
from base64 import b64decode, b64encode
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5 as Cipher_PKCS1_v1_5


class APIRequest:

    def __init__(self, context=None):
        self.context = context

    def execute(self):
        if self.context is not None:
            self.create_default_headers()
            try:
                return {
                    APIMethodType.GET: self.__get,
                    APIMethodType.POST: self.__post,
                    APIMethodType.PUT: self.__put
                }.get(self.context.method_type, self.__unknown)()
            except requests.exceptions.ConnectionError as ce:
                return None
        else:
            raise TypeError('Context cannot be None.')

    def create_bearer_token(self):
        key_der = b64decode(self.context.public_key)
        key_pub = RSA.importKey(key_der)
        cipher = Cipher_PKCS1_v1_5.new(key_pub)
        cipher_text = cipher.encrypt(self.context.api_key.encode('ascii'))
        encrypted_msg = b64encode(cipher_text)

        return encrypted_msg

    def create_default_headers(self):
        self.context.add_header(
            'Authorization',
            'Bearer {}'.format(self.create_bearer_token().decode('utf-8')))
        self.context.add_header('Content-Type', 'application/json')
        self.context.add_header('Host', self.context.address)

    def __get(self):
        r = requests.get(
            self.context.get_url(),
            params=self.context.get_parameters(),
            headers=self.context.get_headers())
        return APIResponse(r.status_code,
                           json.loads(r.headers.__str__().replace("'", '"')),
                           json.loads(r.text))

    def __post(self):
        r = requests.post(self.context.get_url(),
                          headers=self.context.get_headers(),
                          json=self.context.get_parameters())
        print(r)
        return APIResponse(r.status_code,
                           json.loads(r.headers.__str__().replace("'", '"')),
                           json.loads(r.text))

    def __put(self):
        r = requests.put(self.context.get_url(),
                         headers=self.context.get_headers(),
                         json=self.context.get_parameters())
        return APIResponse(r.status_code,
                           json.loads(r.headers.__str__().replace("'", '"')),
                           json.loads(r.text))

    def __unknown(self):
        raise Exception('Unknown Method')


class APIResponse(dict):

    def __init__(self, status_code: str, headers: dict, body: dict):
        super(APIResponse, self).__init__()
        self['status_code'] = status_code
        self['headers'] = headers
        self['body'] = body

    @property
    def status_code(self) -> int:
        return self['status_code']

    @status_code.setter
    def status_code(self, status_code: int):
        if type(status_code) is not int:
            raise TypeError('status_code must be a int')
        else:
            self['status_code'] = status_code

    @property
    def headers(self) -> dict:
        return self['headers']

    @headers.setter
    def headers(self, headers: dict):
        if type(headers) is not dict:
            raise TypeError('headers must be a dict')
        else:
            self['headers'] = headers

    @property
    def body(self) -> dict:
        return self['body']

    @body.setter
    def body(self, body: dict):
        if type(body) is not dict:
            raise TypeError('body must be a dict')
        else:
            self['body'] = body


class APIMethodType(Enum):
    GET: int = 0
    POST: int = 1
    PUT: int = 3
    DELETE: int = 4


class APIContext(dict):

    def __init__(self, api_key: str = '', public_key: str = '',
                 ssl: bool = False, method_type: Enum = APIMethodType.GET,
                 address: str = '', port: int = 80, path: str = '',
                 headers: dict = {}, parameters: dict = {}):
        super(APIContext, self).__init__()

        self['api_key'] = api_key
        self['public_key'] = public_key
        self['ssl'] = ssl
        self['method_type'] = method_type
        self['address'] = address
        self['port'] = port
        self['path'] = path
        self['headers'] = headers
        self['parameters'] = parameters

    def get_url(self):
        if self.ssl is True:
            return 'https://{}:{}{}'.format(self.address, self.port, self.path)
        else:
            return 'http://{}:{}{}'.format(self.address, self.port, self.path)

    def add_header(self, header, value):
        self['headers'].update({header: value})

    def get_headers(self):
        return self['headers']

    def add_parameter(self, key, value):
        self['parameters'].update({key: value})

    def get_parameters(self):
        return self['parameters']

    @property
    def api_key(self) -> str:
        return self['api_key']

    @api_key.setter
    def api_key(self, api_key: str):
        if type(api_key) is not str:
            raise TypeError('api_key must be a str')
        else:
            self['api_key'] = api_key

    @property
    def public_key(self) -> str:
        return self['public_key']

    @public_key.setter
    def public_key(self, public_key: str):
        if type(public_key) is not str:
            raise TypeError('public_key must be a str')
        else:
            self['public_key'] = public_key

    @property
    def ssl(self) -> bool:
        return self['ssl']

    @ssl.setter
    def ssl(self, ssl: bool):
        if type(ssl) is not bool:
            raise TypeError('ssl must be a bool')
        else:
            self['ssl'] = ssl

    @property
    def method_type(self) -> APIMethodType:
        return self['method_type']

    @method_type.setter
    def method_type(self, method_type: APIMethodType):
        if type(method_type) is not APIMethodType:
            raise TypeError('method_type must be a APIMethodType')
        else:
            self['method_type'] = method_type

    @property
    def address(self) -> str:
        return self['address']

    @address.setter
    def address(self, address: str):
        if type(address) is not str:
            raise TypeError('address must be a str')
        else:
            self['address'] = address

    @property
    def port(self) -> int:
        return self['port']

    @port.setter
    def port(self, port: int):
        if type(port) is not int:
            raise TypeError('port must be a int')
        else:
            self['port'] = port

    @property
    def path(self) -> str:
        return self['path']

    @path.setter
    def path(self, path: str):
        if type(path) is not str:
            raise TypeError('path must be a str')
        else:
            self['path'] = path


mobile_payments/constants.py
# BASE_URL = 'openapi.m-pesa.com'
# get_session_url = '/sandbox/ipg/v2/vodacomTZN/getSession/'
# c2bPayment_url = '/sandbox/ipg/v2/vodacomTZN/c2bPayment/singleStage/'
# reversal_url = '/sandbox/ipg/v2/vodacomTZN/reversal/'
# b2cPayment_url = '/sandbox/ipg/v2/vodacomTZN/b2cPayment/'
# b2bPayment_url = '/openapi/ipg/v2/vodacomTZN/b2bPayment/'
# transaction_status_url = '/openapi/ipg/v2/vodacomTZN/queryTransactionStatus/'

MPESA_PUBLIC_KEY = 'MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEArv9yxA69XQKBo24BaF/D+fvlqmGdYjqLQ5WtNBb5tquqGvAvG3WMFETVUSow/LizQalxj2ElMVrUmzu5mGGkxK08bWEXF7a1DEvtVJs6nppIlFJc2SnrU14AOrIrB28ogm58JjAl5BOQawOXD5dfSk7MaAA82pVHoIqEu0FxA8BOKU+RGTihRU+ptw1j4bsAJYiPbSX6i71gfPvwHPYamM0bfI4CmlsUUR3KvCG24rB6FNPcRBhM3jDuv8ae2kC33w9hEq8qNB55uw51vK7hyXoAa+U7IqP1y6nBdlN25gkxEA8yrsl1678cspeXr+3ciRyqoRgj9RD/ONbJhhxFvt1cLBh+qwK2eqISfBb06eRnNeC71oBokDm3zyCnkOtMDGl7IvnMfZfEPFCfg5QgJVk1msPpRvQxmEsrX9MQRyFVzgy2CWNIb7c+jPapyrNwoUbANlN8adU1m6yOuoX7F49x+OjiG2se0EJ6nafeKUXw/+hiJZvELUYgzKUtMAZVTNZfT8jjb58j8GVtuS+6TM2AutbejaCV84ZK58E2CRJqhmjQibEUO6KPdD7oTlEkFy52Y1uOOBXgYpqMzufNPmfdqqqSM4dU70PO8ogyKGiLAIxCetMjjm6FCMEA3Kc8K0Ig7/XtFm9By6VxTJK1Mg36TlHaZKP6VzVLXMtesJECAwEAAQ =='

MPESA_API_KEY = '917qPI7PfSQNaLBE6Gr9GYNzVWH0pwOC'

MPESA_BASE_URL = 'openapi.m-pesa.com'
MPESA_GET_SESSION_URL = '/sandbox/ipg/v2/vodacomTZN/getSession/'
MPESA_C2BPAYMENT_URL = '/sandbox/ipg/v2/vodacomTZN/c2bPayment/singleStage/'
MPESA_REVERSAL_URL = '/sandbox/ipg/v2/vodacomTZN/reversal/'
MPESA_B2CPAYMENT_URL = '/sandbox/ipg/v2/vodacomTZN/b2cPayment/'
MPESA_B2BPAYMENT_URL = '/openapi/ipg/v2/vodacomTZN/b2bPayment/'
MPESA_TRANSACTION_STATUS_URL = '/openapi/ipg/v2/vodacomTZN/queryTransactionStatus/'


mobile_payments/vodacom.py
"""
vodacom.py
-----------
This module contains the class for all M-Pesa transactions.
"""

from .constants import *
from .open_api import APIContext, APIMethodType, APIRequest

BASE_URL = MPESA_BASE_URL
get_session_url = MPESA_GET_SESSION_URL
c2bPayment_url = MPESA_C2BPAYMENT_URL
reversal_url = MPESA_REVERSAL_URL
b2cPayment_url = MPESA_B2CPAYMENT_URL
b2bPayment_url = MPESA_B2BPAYMENT_URL
transaction_status_url = MPESA_TRANSACTION_STATUS_URL


class MPESA:

    def __init__(self, api_key: str, public_key: str,
                 ssl: bool = True) -> None:
        """Generate context required for making transaction

        :param api_key: API key for your application
        :type api_key: string
        :param public_key: Open API public key
        :type public_key: string
        :param ssl: Either to use ssl or not, defaults to True
        :type ssl: bool, optional
        """
        self.context = APIContext(
            api_key, public_key, ssl=ssl, address=BASE_URL, port=443)
        self.context.add_header('Origin', '*')

    def get_encrypted_api_key(self) -> str:
        """A function to return encrypted API key

        :return: Encrypted API key
        :rtype: str
        """
        return APIRequest(self.context).create_bearer_token()

    def get_session_id(self, path: str = get_session_url) -> str:
        """A function to generate valid Session ID needed to transact on M-Pesa
        using OpenAPI.

        :param path: url, defaults to get_session_url
        :type path: string, optional
        :raises Exception: When request fails, exception must be raised.
        :return: A valid Session ID
        :rtype: str
        """
        self.context.update({'method_type': APIMethodType.GET,
                             'path': path})

        response = None

        try:
            response = APIRequest(self.context).execute()
        except Exception as e:
            print('Call Failed: ', e)

        if response is None:
            raise Exception(
                'SessionKey call failed to get response. Please check.')
        else:
            return response.body['output_SessionID']

    def _get_api_response(self, context: dict) -> dict:
        """A function for getting results from API call.

        :param context: A dictionary containing all the necessary
        parameters for making API call.
        :type context: dict
        :raises Exception: Exception raised when API call fails.
        :return: Response from API call.
        :rtype: dict
        """
        response = None
        try:
            response = APIRequest(context).execute()
        except Exception as e:
            print('Call Failed: ', e)

        if response is None:
            raise Exception('API call failed to get result. Please check.')
        else:
            return response

    def customer2business(self, parameters: dict,
                          path: str = c2bPayment_url) -> dict:
        """A standard customer-to-business transaction

        :param parameters: A dictionary containing all necessary
        key value pairs.
        :type parameters: dict
        :param path: url for customer-to-business transaction,
        defaults to c2bPayment_url
        :type path: str, optional
        :return: Response from API call.
        :rtype: dict

        Example of parameters:

        parameters = {
            'input_Amount': 10,
            'input_Country': 'TZN',
            'input_Currency': 'TZS',
            'input_CustomerMSISDN': '000000000001',
            'input_ServiceProviderCode': '000000',
            'input_ThirdPartyConversationID':
                'asv02e5958774f7ba228d83d0d689761',
            'input_TransactionReference': 'T1234C',
            'input_PurchasedItemsDesc': 'Shoes',
        }
        """

        self.context.update({
            'api_key': self.get_session_id(),
            'method_type': APIMethodType.POST,
            'path': path,
            'parameters': {k: v for k, v in parameters.items()}
        })

        response = self._get_api_response(self.context)
        return response

    def reversal(self, reversal_parameters: dict,
                 path: str = reversal_url) -> dict:
        """Reverse a successful transaction.

        :param reversal_parameters: A dictionary containing all the
        necessary information for reversing transaction.
        :type reversal_parameters: dict
        :param path: url for reversing transaction, defaults to reversal_url
        :type path: str, optional
        :return: Dictionary of reversed transaction when successful.
        :rtype: dict

        Example of reversal_parameters:

        reversal_parameters = {
            'input_ReversalAmount': '25',
            'input_Country': 'TZN',
            'input_ServiceProviderCode': '000000',
            'input_ThirdPartyConversationID':
                'asv02e5958774f7ba228d83d0d689761',
            'input_TransactionID': '0000000000001',
        }
        """
        self.context.update({
            'api_key': self.get_session_id(),
            'method_type': APIMethodType.PUT,
            'path': path,
            'parameters': {k: v for k, v in reversal_parameters.items()}
        })

        response = self._get_api_response(self.context)
        return response

    def business2customer(self, parameters: dict,
                          path: str = b2cPayment_url) -> dict:
        """A standard customer-to-business transaction.

        :param parameters: Information required for successful transaction.
        :type parameters: dict
        :param path: url for business to customer payment,
        defaults to b2cPayment_url
        :type path: str, optional
        :return: Response from API call.
        :rtype: dict

        Example of paramters:

        parameters = {
            'input_Amount': '10',
            'input_Country': 'TZN',
            'input_Currency': 'TZS',
            'input_CustomerMSISDN': '000000000001',
            'input_ServiceProviderCode': '000000',
            'input_ThirdPartyConversationID':
            'asv02e5958774f7ba228d83d0d689761',
            'input_TransactionReference': 'T1234C',
            'input_PaymentItemsDesc': 'Salary payment',
        }
       """
        self.context.update({
            'api_key': self.get_session_id(),
            'method_type': APIMethodType.POST,
            'path': path,
            'parameters': {k: v for k, v in parameters.items()}
        })

        response = self._get_api_response(self.context)
        return response

    def business2business(self, parameters: dict,
                          path: str = b2bPayment_url) -> dict:
        """Business-to-business transactions (Single Stage).

        :param parameters: Information necessary for business-to-business
        transaction.
        :type parameters: dict
        :param path: url for business-to-business transaction,
        defaults to b2bPayment_url
        :type path: str, optional
        :return: Response from API call.
        :rtype: dict

        Example of parameters:

        parameters = {
            'input_Amount': '10',
            'input_Country': 'TZN',
            'input_Currency': 'TZS',
            'input_CustomerMSISDN': '000000000001',
            'input_ServiceProviderCode': '000000',
            'input_ThirdPartyConversationID':'asv02e5958774f7ba228d83d0d689761',
            'input_TransactionReference': 'T1234C',
            'input_PaymentItemsDesc': 'Salary payment',
            }
        """
        self.context.update({
            'api_key': self.get_session_id(),
            'method_type': APIMethodType.POST,
            'path': path,
            'parameters': {k: v for k, v in parameters.items()}
        })

        response = self._get_api_response(self.context)
        return response

    def status(self, parameters: dict,
               path: str = transaction_status_url) -> dict:
        """Query the status of the transaction that has been initiated.

        :param parameters: Information necessary for querying
        transaction status.
        :type parameters: dict
        :param path: url for querying transaction status,
        defaults to transaction_status_url
        :type path: str, optional
        :return: Response from API call.
        :rtype: dict

        Example of paramters:

        parameters = {
            'input_QueryReference': '000000000000000000001',
            'input_ServiceProviderCode': '000000',
            'input_ThirdPartyConversationID':'asv02e5958774f7ba228d83d0d689761',
            'input_Country': 'TZN',
        }
        """
        self.context.update({
            'api_key': self.get_session_id(),
            'method_type': APIMethodType.GET,
            'path': path,
            'parameters': {k: v for k, v in parameters.items()}
        })

        response = self._get_api_response(self.context)
        return response
